### ISSUE TITLE
PARAFAC update operation (MTTKRP)

### ISSUE THREAD

**Issue by dawonahn** [tensorly/tensorly#210]
Good evening.
#### Describe the bug

I found a weird thing in the MTTKRP operation.
The order of factor matrices in MTTKRP is not the same with a formulation of CP-ALS.
However, the fit (factorization result) of tensorly version is way better than the fit of my version (implementing the same formulation of CP-ALS). 
I referred the [Kolda's survey paper](https://www.kolda.net/publication/TensorReview.pdf).

Last but not least, tensorly library is very useful for my study and research.
I thank you for your contribution to tensor learning tools.
 
#### Steps or Code to Reproduce

Short version of parafac code in tensorly is below:

```python
for mode in modes_list:
   pseudo_inverse = tl.tensor(np.ones((rank, rank)), **tl.context(tensor))
   for i, factor in enumerate(factors):
     if i != mode:
       pseudo_inverse = pseudo_inverse*tl.dot(tl.conj(tl.transpose(factor)), factor)
       
    mttkrp = unfolding_dot_khatri_rao(tensor, (None, factors), mode)

    factor = tl.transpose(tl.solve(tl.conj(tl.transpose(pseudo_inverse)),
                                    tl.transpose(mttkrp)))
    factors[mode] = factor
```

In the codes, there is a code about mttkrp (See the 7th line).
According to CP-ALS formulation, the order of factor matrices in `factors` has to be reversed.

For example, 
when updating the first factor matrix (A) of three order tensor factorization, 
the update will be like this: Aˆ = X(1)(C ⊙ B)(CTC ∗ BTB)†.
**Therefore the mttkrp should be X(1)(C ⊙ B) while the tensorly code indicates like Aˆ = X(1)(B ⊙ C).**

#### Actual result
However the performance of my version is very poor.

Performances of synthetic tensor are below:
tensorly cp: 0.4406
my cp: 0.0421

##### Tensorly version ( I implemented unfolding_dot_khatri_rao with tensorly API)

```python
    A = factors[0]
    B = factors[1]
    C = factors[2]
    for i in range(n_iter_max):
        V = tl.dot(tl.transpose(C), C)*tl.dot(tl.transpose(B), B)
        mttkrp = tl.dot(tl.unfold(tensor, 0), tl.tenalg.khatri_rao([B, C]))
        A = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))
        
        V = tl.dot(tl.transpose(C), C)*tl.dot(tl.transpose(A), A)
        mttkrp = tl.dot(tl.unfold(tensor, 1), tl.tenalg.khatri_rao([A, C]))
        B = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))
        
        V = tl.dot(tl.transpose(B), B)*tl.dot(tl.transpose(A), A)
        mttkrp = tl.dot(tl.unfold(tensor, 2), tl.tenalg.khatri_rao([A, B]))
        C = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))

    return KruskalTensor((None, [A, B, C]))
```

##### My edited version
```python
    A = factors[0]
    B = factors[1]
    C = factors[2]
    for i in range(n_iter_max):
        V = tl.dot(tl.transpose(C), C)*tl.dot(tl.transpose(B), B)
        mttkrp = tl.dot(tl.unfold(tensor, 0), tl.tenalg.khatri_rao([C, B]))
        A = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))
        
        V = tl.dot(tl.transpose(C), C)*tl.dot(tl.transpose(A), A)
        mttkrp = tl.dot(tl.unfold(tensor, 1), tl.tenalg.khatri_rao([C, A]))
        B = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))
        
        V = tl.dot(tl.transpose(B), B)*tl.dot(tl.transpose(A), A)
        mttkrp = tl.dot(tl.unfold(tensor, 2), tl.tenalg.khatri_rao([B, A]))
        C = tl.transpose(tl.solve(tl.transpose(V), tl.transpose(mttkrp)))
    
    return KruskalTensor((None, [A, B, C]))
```

#### Versions
Linux-4.4.0-178-generic-x86_64-with-debian-stretch-sid
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0]
NumPy 1.16.2
SciPy 1.4.1
TensorLy 0.4.5

**Comment by lucetre** [comment]
I'm having the same problem. Happy to find this issue ;)

**Comment by JeanKossaifi** [comment]
Glad you're finding TensorLy useful for your studies and research!

The difference you're seeing comes from the fact that we use a different definition of unfolding that leads to better performance particularly on GPU (more generally when using C-ordering of the elements).

You can check the definition in the [documentation](http://tensorly.org/stable/user_guide/tensor_basics.html). There is more background in this [blog post](http://jeankossaifi.com/blog/unfolding.html) and this [Stack Overflow](https://stackoverflow.com/a/50086319/4802196) answer.

In short, the unfolding we use differs by the ordering of the fibers, which mirror the natural order that appears in Kronecker / Khatri-Rao product. As a result, you no longer need to reverse the order of the factors you're taking the product with.

**Comment by dawonahn** [comment]
Thank you for your kind reply and for sharing your blog post! That's very informative 👍 
I will scrutinize codes again before opening issues.
